### PR TITLE
Remove RL use of mock dataloader and kill RL inference interface on exit

### PR DESCRIPTION
**Pull request by jon-barker**
# What does this PR do ?

Prevents hanging on exit of the pretrain function during RL by:

1. Removing the use of a mock LM dataloader when running the RL loop. 
2. Ensuring the RL inference interfaces are killed on exit.

UPDATE:  i have confirmed that we can run a qwen2p5_3b rl training workload including checkpointing and resume and with evaluation.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/training.py [hunk 1]
```diff
@@ -833,6 +833,9 @@ def pretrain(
         {'app_finish_time': one_logger_utils.get_timestamp_in_ms()}
     )
 
+    if getattr(args, 'perform_rl_step', False):
+        rl_utils.rl_inference_interface_shutdown()
```
Comment (yobibyte) [inline megatron/training/training.py:837]: why do we do getattr instead of args.perform_rl_step?

File: megatron/training/training.py [hunk 2]
```diff
@@ -2859,33 +2863,38 @@ def build_train_valid_test_data_loaders(build_train_valid_test_datasets_provider
...
-        # Build dataloders.
-        train_dataloader = build_pretraining_data_loader(train_ds, args.consumed_train_samples)
+        # Build datasets and dataloders.
+        if getattr(args, 'perform_rl_step', True):
+            # we don't need to build any dataloaders for RL training
+            train_dataloader = None
+            valid_dataloaders = None
+            test_dataloader = None
+            do_train = args.train_iters > 0
+            do_valid = False
```
Comment (yobibyte) [inline megatron/training/training.py]: Won't this skip validation phase? We use validation in rl runs.

File: megatron/training/training.py [hunk 3]
```diff
@@ -2859,33 +2863,38 @@ def build_train_valid_test_data_loaders(build_train_valid_test_datasets_provider
...
+            do_test = False
+        else:
+            train_ds, valid_ds, test_ds = build_train_valid_test_datasets(
+                build_train_valid_test_datasets_provider, (1, 1, 1) if getattr(args, 'perform_rl_step', False) else None
```
Comment (yobibyte) [inline megatron/training/training.py]: We can skip the getattr check here as we are in the else branch of checking it for being True.

File: megatron/training/training.py (hunk 1 above)
Comment (jon-barker) [reply megatron/training/training.py:837]: Consistency with the rest of the file. I agree that `args.perform_rl_step` is better but we should change that everywhere in a subsequent PR IMO

File: megatron/training/training.py (hunk 2 above)
Comment (jon-barker) [reply megatron/training/training.py]: I think you’re right. I’ll fix and test validation works. Thanks

**Comment by jon-barker** [comment]
> Hey @jon-barker , thanks for adding the fix.
> 
> Re skipping building the mock dataset, I think we will stop running evaluations in the current setup. Could you double check, please?

@yobibyte can you please re-review - I've confirmed that with the latest commit evaluation runs and also checkpointing for both save and load is working fine without a mock dataloader. So i think this is good to go.

File: megatron/training/training.py (hunk 3 above)
Comment (jon-barker) [reply megatron/training/training.py]: done

**Review by yobibyte — APPROVED** [review]
